src/app/core/services/scheduling.service.ts

```diff
@@ -209,6 +222,44 @@ export class SchedulingService {
     })
   }
 
+  addToCompletedTasks(task) {
+    return this.storage.push(StorageKeys.SCHEDULE_TASKS_COMPLETED, task)
+  }
+
+  updateScheduleWithCompletedTasks(schedule) {
+    // NOTE: If utcOffsetPrev exists, timezone has changed
+    if (this.utcOffsetPrev) {
+      const currentMidnight = new Date().setHours(0, 0, 0, 0)
+      const prevMidnight =
+        new Date().setUTCHours(0, 0, 0, 0) + this.utcOffsetPrev * 60000
+      this.completedTasks.map(d => {
+        const index = schedule.findIndex(
+          s =>
+            s.timestamp - currentMidnight == d.timestamp - prevMidnight &&
+            s.name == d.name
+        )
+        if (index > -1) {
+          schedule[index].completed = true
+          return this.addToCompletedTasks(schedule[index])
+        }
+      })
+    } else {
+      this.completedTasks.map(d => {
+        if (
+          schedule[d.index].timestamp == d.timestamp &&
+          schedule[d.index].name == d.name
+        ) {
+          schedule[d.index].completed = true
+          return this.addToCompletedTasks(schedule[d.index])
+        }
+      })
+    }
+    this.storage.remove(StorageKeys.UTC_OFFSET_PREV)
+    this.storage.remove(StorageKeys.SCHEDULE_TASKS_COMPLETED)
```

Comment:
Why are we even using a storage key here ? From what i can see there is no need for storing the completed tasks in storage as they are calculated using the existing schedule, right ?

Reply:
For when a protocol is changed?

Reply:
Yes but the schedule already has information of completed tasks. So why are we storing these in a separate key ?

Reply:
Okay we can merge for now. We can revisit this later when we refactor the code.